docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+company's running containers. Even though his company has no standard base
+image, he's able to audit all containers using:
+
+```
+% for pod in $(kubectl get -o name pod); do
+    kubectl debug -m gcr.io/neato/security-audit -p $pod /security-audit.sh
+  done
+```
+
+## Alternatives Considered
+
+### Mutable Pod Spec
+
+Rather than adding an operation to have Kubernetes attach a pod we could instead
+make the pod spec mutable so the client can generate an update adding a
+container. `SyncPod()` has no issues adding the container to the pod at that
+point, but an immutable pod spec has been a basic assumption in Kubernetes thus
+far and changing it carries risk. It's preferable to keep the pod spec immutable
+as a best practice.
+
+### Ephemeral container
+
+An earlier version of this proposal suggested running an ephemeral container in
+the pod namespaces. The container would not be added to the pod spec and would
+exist only as long as the process it ran. This has the advantage of behaving
+similarly to the current kubectl exec, but it is opaque and likely violates
+design assumptions. We could add constructs to track and report on both
+traditional exec process and exec containers, but this would probably be more
+work than adding to the pod spec. Both are generally useful, and neither
+precludes the other in the future, so we chose mutating the pod spec for
+expedience.
+
+### Attaching Container Type Volume
+
+Combining container volumes ([#831](https://issues.k8s.io/831)) with the ability
+to add volumes to the pod spec would get us most of the way there. One could
+mount a volume of debug utilities at debug time. Docker does not allow adding a
+volume to a running container, however, so this would require a container
+restart. A restart doesn't meet our requirements for troubleshooting.
+
+Rather than attaching the container at debug time, kubernetes could always
+attach a volume at a random path at run time, just in case it's needed. Though
+this simplifies the solution by working within the existing constraints of
+`kubectl exec`, it has a sufficient list of minor limitations (detailed in
+[#10834](https://issues.k8s.io/10834)) to result in a poor user experience.
+
+### Inactive container
+
+If Kubernetes supported the concept of an "inactive" container, we could
+configure it as part of a pod and activate it at debug time. In order to avoid
+coupling the debug tool versions with those of the running containers, we would
+need to ensure the debug image was pulled at debug time. The container could
+then be run with a TTY and attached using kubectl. We would need to figure out a
+solution that allows access the filesystem of other containers.
+
+The downside of this approach is that it requires prior configuration. In
+addition to requiring prior consideration, it would increase boilerplate config.
+A requirement for prior configuration makes it feel like a workaround rather
+than a feature of the platform.
+
+### Implicit Empty Volume
```

Comment:
I definitely want the thing that feels most natural. For me that's running a container since that's what we do everywhere else in Kubernetes. We by necessity have an infrastructure for distributing container images so I'd prefer to use that rather than have the kubelet/runtime unpacking tarballs, though it might be possible to do something like union mount a container image at debug time.

Adding a container doesn't require any additional complexity on the run times, which is nice given the diversity we want to support.

Making this work with the existing `kubectl exec` (by adding binaries) makes a lot of other things easy, but I don't like surfacing that complexity to the user. Let's say we mount debian's userland at /adc2ce42-e1af-11e6-bf01-fe55135034f3. kubectl can figure out how to exec /bin/bash by looking up its mount point, but bash has no way to figure out that curl is at /adc2ce42-e1af-11e6-bf01-fe55135034f3/usr/bin. We can't set env variables for exec, so we need to embed a launcher within the image. kubectl calls `/adc2ce42-e1af-11e6-bf01-fe55135034f3/launch bash` which sets PATH=/adc2ce42-e1af-11e6-bf01-fe55135034f3/bin and exec()s bash. If the user wants to know what utilities are installed, `ls /bin` will be very confusing. Instead the user should `which bash` or `ls /` and then `ls /adc2ce42-e1af-11e6-bf01-fe55135034f3`. If a tool is missing, `apt-get` won't work, you have to mount a different image, which might unpack tools over the top of your mount point.

The thing I like about running a new container is that it behaves like a regular Linux userland.